web/tests/lib/test_stream.js

```diff
@@ -0,0 +1,46 @@
+"use strict";
+
+let stream_id = 1;
+
+const get_stream_id = () => {
+    stream_id += 1 + Math.floor(Math.random() * 10);
+    return stream_id;
+};
```

Comment:
I'm not I see the benefit of having any randomness here; that seems like it'll take deterministic bugs related to reused stream IDs and make them into annoying to investigate nondeterministic bugs.

I think it'd reasonable to have the first ID be something like 20000, and use a different first digit for users/groups/etc., in order to make the type of an ID guessable, but just incrementing seems more natural than the randomness.

Reply:
FWIW I don't think we've ever seen any flakes in zulip-mobile or zulip-flutter related to this sort of randomly-chosen ID, and we've had it there since 2019 (zulip/zulip-mobile@abe2d662fcc8c5e0acf6b97070541b8e5c68cd5f).

Or more precisely: we had a flake or two that was related to random IDs colliding, and a flake or two from them coming in different orders on different runs. We fixed it so they never collided, then made them come in increasing order; and since then we've had no other flakes.

The reason I think this randomization is helpful is that it prevents writing tests that use a generated stream in one place, and then in another place assume its stream ID was 20007 or some such value. That will work deterministically, but be very brittle: when someone edits some other test in the file so that there's one more or fewer stream generated, a bunch of arbitrary-looking numbers will have to get updated.

Conversely, as I wrote in the original commit message:

    (General point: randomness in tests is bad if it makes a test
    occasionally fail, but fine if the test almost always fails.
    I'm hoping this will mainly be the latter.)

Having the IDs collide, or come sometimes in one order and sometimes another, can basically be seen as ways that the original statelessly-random version would cause tests to sometimes fail, which is why we had to fix those. When they're random but in increasing order, those tests that hardcode 20007 or whatever will almost always fail, which is fine; and empirically they no longer seem to produce tests that occasionally fail.

Reply:
Yeah, maybe the randomness is good. I maybe what I objected to was more randomness with the numbers starting from `1` as a setup.

Reply:
Yeah, definitely it's important for the randomly-generated values to come from a range of large numbers (for example, starting at 20000) where it'd be unlikely someone would happen to make up the same number in hard-coded data. If they started at 1, then that'd be a recipe for tests that occasionally fail because someone would have a hard-coded stream ID like 1 or 3 or 10 and the random ID would occasionally collide with it.